### PR TITLE
Upgrading to React Router V4

### PR DESCRIPTION
**Pull request by nshntarora**

### CODE REVIEW

**Comment by nshntarora** [comment]
@krishnagopinath I believe I'm done here. Could you please test and let me know if something's broken? If not, let's merge it.

**Comment by krishnagopinath** [comment]
Whoohoo! I'll get to this by the evening :D (In the middle of moving to a new house)

**Comment by krishnagopinath** [comment]
Everything seems OK, but I think we need to re-do our auth model a bit, which we could once I finish up the backend thing. Since we're going to change how our onboarding process works (hopefully with emails, clean setup, etc.), we should probably revisit how we use the router for accessing routes that need auth. We could make a `PrivateRoute` HOC that we could use in the router to protect authenticated routes, but not for right now.

File: package.json [hunk 1]
```diff
@@ -30,6 +30,7 @@
   "homepage": "http://matterwiki.com",
   "dependencies": {
     "axios": "^0.16.1",
+    "babel-helpers": "^6.24.1",
```
Comment (krishnagopinath) [inline package.json]: We'll have to remove this 😸

File: src/client/components/Layout/components/Header/components/AppNavbar/components/SearchForm.js [hunk 2]
```diff
@@ -1,6 +1,6 @@
 import React from "react";
 import { Navbar, Form, FormGroup, FormControl, Button } from "react-bootstrap";
-import { hashHistory } from "react-router";
+// import { hashHistory } from "react-router";
```
Comment (krishnagopinath) [inline src/client/components/Layout/components/Header/components/AppNavbar/components/SearchForm.js]: If we're moving away take this comment away wont you :D

File: src/client/index.js [hunk 3]
```diff
@@ -1,28 +1,30 @@
...
+renderApp();
 
 // react HMR
-if(module.hot) {
-  module.hot.accept('./routes.js', () => {
-    const routeChanges = require('./routes.js').default;
-    renderApp(routeChanges);
-  })
-}
+// if(module.hot) {
```
Comment (krishnagopinath) [inline src/client/index.js]: is HMR screwed up now? :(

File: src/client/scenes/Admin/Admin.js [hunk 4]
```diff
@@ -25,18 +32,26 @@ class Admin extends React.Component {
...
+    }
   };
 
   render() {
     return (
       (this.state.loading && <Loader />) ||
       <div>
         <AdminNavBar handleSelect={this.updateTab} activeTab={this.state.tab} />
-        {this.props.children}
+        {/*{this.props.children}*/}
```
Comment (krishnagopinath) [inline src/client/scenes/Admin/Admin.js]: Reminder to remove this in case you forget

File: src/client/scenes/Admin/Admin.js [hunk 5]
```diff
@@ -25,18 +32,26 @@ class Admin extends React.Component {
...
+        { (this.props.location.pathname === '/admin') ? <Redirect to={`${this.props.match.url}/users`} /> : ""}
```
Comment (krishnagopinath) [inline src/client/scenes/Admin/Admin.js]: You could save yourself SO many keystrokes by picking all this from `this`, before hand:

```
render() {
    const {location, match} = this.props;
    // ....
    <Route path={`${match.url}/topics`} component={ManageTopics} />
   // ....
}
```

File: src/client/scenes/Admin/scenes/Topics.js [hunk 6]
```diff
@@ -1,9 +1,10 @@
 import React from "react";
-import { hashHistory, Link } from "react-router";
+// import { hashHistory, Link } from "react-router";
```
Comment (krishnagopinath) [inline src/client/scenes/Admin/scenes/Topics.js]: beep beep. comment removal reminder

File: src/client/scenes/Admin/scenes/Users.js [hunk 7]
```diff
@@ -1,9 +1,10 @@
 import React from "react";
-import { hashHistory, Link } from "react-router";
+// import { hashHistory, Link } from "react-router";
```
Comment (krishnagopinath) [inline src/client/scenes/Admin/scenes/Users.js]: beep beep. comment removal reminder

File: src/client/scenes/Article/Article.js [hunk 8]
```diff
@@ -0,0 +1,23 @@
...
+import {
+  NewArticle,
+  ViewArticle,
+  EditArticle,
+  Archives
+} from "./scenes/AllScenesArticle";
+
+const Article = ({match}) =>
+  <div>
+    {/*{children}*/}
```
Comment (krishnagopinath) [inline src/client/scenes/Article/Article.js]: beep beep. comment removal reminder

File: src/client/scenes/Errors/404.js [hunk 9]
```diff
@@ -0,0 +1,24 @@
+import React from "react";
```
Comment (krishnagopinath) [inline src/client/scenes/Errors/404.js]: <3

Comment (krishnagopinath) [inline src/client/scenes/Errors/404.js]: Very nice. Finally a 404 page :D

File: src/client/scenes/Main/AllScenesMain.js [hunk 10]
```diff
@@ -0,0 +1,9 @@
+import Login from "../Login/Login";
```
Comment (krishnagopinath) [inline src/client/scenes/Main/AllScenesMain.js]: A reminder to rename these to `MainScenes`

**Comment by nshntarora** [comment]
Made all the changes you suggested :smile: Ready to merge

**Comment by krishnagopinath** [comment]
Merge it in, sir! :shipit: 

Will merge the server next and then the ESLint stuff. Will need you to check if everything works as intended! :D

**Pull request merged**